File: source/class/ship/playership.cpp
/*
	playership.cpp
	Creado por Rafael Alcalde Azpiazu (nekerafa@gmail.com) el vie 25 dic 2015 00:39:21 (CET)

	Cabecera de la clase nave del jugador.

	Space Cats is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	Space Cats is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
*/


#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include <SDL2/SDL_ttf.h>
#include <string>
#include <algorithm>
#include "../../include/window.h"
#include "../../include/texture.h"
#include "../../include/timer.h"
#include "../../include/weapon.h"
#include "../../include/ship/ship.hpp"
#include "../../include/ship/playership.hpp"

extern texture * img_bullet;

// Crea una nave
playership::playership(int x, int y, int life) : ship(x, y, life) {
	textureship->load("./res/ships/ship-cat.png");
	textureboost->load("./res/ships/combustion.png");
	textureboom->load("./res/ships/explosion.png");
	weaponship = new weapon(img_bullet, 250);
}

// Destruye la nave
playership::~playership() {}

// Actualiza los elementos internos
void playership::update() {
	frameboost = std::min((int) timerboost->time()/100, 3);
	if (timerboost->time() >= 300) { timerboost->reset(); timerboost->start(); }
	if (life == 0 && timerboom->time() == 0) timerboom->start();
	frameboom = std::min((int) timerboom->time()/100, 7);
	if (timerboom->time() >= 700 && !kill) { kill = true; }
}

// Renderiza la nave
void playership::render() {
	if(life > 0) {
		textureship->render(x-16, y-16);
		textureboost->render(x-32, y-8, 0, frameboost*16, 16, 16);
	} else textureboom->render(x-16, y-16, 0, frameboom*32, 32, 32);
	weaponship->render();
}

// Mueve la nave
void playership::move() {
	const Uint8 *key = SDL_GetKeyboardState(NULL);
	
	if (key[SDL_SCANCODE_W] && y > 45) y -= 2;
	if (key[SDL_SCANCODE_S] && y < SCREEN_HEIGHT-16) y += 2;
	if (key[SDL_SCANCODE_A] && x > 16) x -= 2;
	if (key[SDL_SCANCODE_D] && x < SCREEN_WIDTH-16) x += 2;
	if (key[SDL_SCANCODE_SPACE]) weaponship->shoot(x+16, y, 4);
}

// Obtiene la hit box de la nave
SDL_Rect playership::getHitBox() {
	SDL_Rect hit_box = {x-16, y-16, 32, 32};
	return hit_box;
}

File: löve source/objects/playership.lua
--[[
    * playership.lua
	* Nave del jugador, es un gatico cute
    * Creado por Rafael Alcalde Azpiazu el jue 23 jun 2016 18:54 (CEST)

	Space Cats es un juego Shoot'em up basado en una guerra espacial de gatos.

	Space Cats is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	Space Cats is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with Foobar. If not, see <http://www.gnu.org/licenses/>.
]]

if not ship then require "objects/ship" end
local weapon = require "require/weapon"
local anim = require "require/sodapop"

-- Modulo nave
playership = {}
 
local ship_cat = love.graphics.newImage("images/ships/ship-cat.png")
local ship_flame = love.graphics.newImage("images/ships/combustion.png")
ship_cat:setFilter("nearest")
ship_flame:setFilter("nearest")

-- Prototipo de nave
function playership.new(x, y, life)
    -- Variables y métodos privados
    local ship = ship.new(x, y, life) -- Nave
	local invulnerabity = timer.new() -- Establece un momento de invulnerabilidad
	local blended = false -- Si está con el shader blend
	local flame = anim.newAnimatedSprite(0, 0) -- Animación del fuego del motor
	local weapon = weapon.new(250) -- Arma
	
	flame:addAnimation('run', {
		image       = ship_flame,
		frameWidth  = 16,
		frameHeight = 16,
		frames      = {
			{1, 1, 1, 4, 0.05}
		}
	})

    -- Variables y métodos públicos
    public = {}
    
	-- Obtiene la posición X
    function public:getX()
        return math.round(ship.x)
    end
    
	-- Obtiene la posición Y
    function public:getY()
        return math.round(ship.y)
    end
    
	-- Obtiene la vida
    function public:getLife()
        return ship.life
    end
    
	-- Aplica daño a la nave
    function public:makeDamage()
		if ship.life > 0 and invulnerabity:getTime() == 0 then
			ship.life = ship.life - 1
			invulnerabity:start(2000)
		end
	end
	
	-- Devuelve verdadero si está destruida la nave
	function public:isKilled()
		return ship.life == 0
	end
	
	-- Función para actualizar los elementos internos
	function public:update(dt)
		if invulnerabity:isFinished() then
			invulnerabity:stop()
			flame.color = {255, 255, 255, 255}
		end
		flame:update(dt)
		weapon:update(dt)
	end
	
	-- Función para rederizar la nave
	function public:render()
		local r, g, b, a = love.graphics.getColor()
		weapon:render()
		-- Ajustamos primero los shaders
		if invulnerabity:getTime() > 0 then
			love.graphics.setColor(255, 255, 255, 255-math.abs((invulnerabity:getTime()%511)-255))
			flame.color = {255, 255, 255, 255-math.abs((invulnerabity:getTime()%511)-255)}
		end
		-- Mostramos la nave y el fuego
		love.graphics.draw(ship_cat, ship.x, ship.y, 0, 1, 1, 16, 16)
		flame:draw(ship.x-24, ship.y)
		love.graphics.setColor(r, g, b, a)
	end

	-- Función para mover la nave
	function public:move(dt)
		-- Movimiento de la nave
		if love.keyboard.isDown("up") then
			ship.y = ship.y - 150*dt
		elseif love.keyboard.isDown("down") then
			ship.y = ship.y + 150*dt
		elseif love.keyboard.isDown("left") then
			ship.x = ship.x - 150*dt
		elseif love.keyboard.isDown("right") then
			ship.x = ship.x + 150*dt
		end
		
		-- Dispara la nave
		if love.keyboard.isDown("space") then
			weapon:shoot(ship.x+16, ship.y, 4)
		end
		
		-- Para que no se salga de la pantalla
		ship.y = math.max(16, math.min(love.game.getHeight()-16, ship.y))
		ship.x = math.max(16, math.min(love.game.getWidth()-16, ship.x))
	end

	-- Obtiene la hitbox de la nave
	function public:getHitBox()
		return {x = ship.x-ship_cat:getWidth()/2,
			    y = ship.y-ship_cat:getHeight()/2,
				w = ship_cat:getWidth(),
				h = ship_cat:getHeight()}
	end
	
	-- Funciones para las balas
	public.bullet = {}

	-- Obtiene la hitbox de la bala
	function public.bullet:getHitBox(i)
		return weapon:getHitBox(i)
	end
	
	-- Destruye la bala
	function public.bullet:kill(i)
		weapon:kill(i)
	end
	
	-- Cantidad de balas disparadas
	function public.bullet:shotBullets()
		return weapon:shotBullets()
	end
	
	-- Libera la nave y sus variables
	function public:free()
		invulnerabity:free()
		invulnerabity = nil
		flame = nil
		weapon:free()
		weapon = nil
		ship:free()
		ship = nil
		public.bullet = nil
		public = nil
		collectgarbage('collect')
		return true
	end
	
    return public
end

function playership.free()
	ship_cat = nil
	ship_flame = nil
	weapon = nil
	anim = nil
	collectgarbage('collect')
	return true
end